File: bot/start/router.py
import re

from aiogram import types, Router
from aiogram.filters import CommandStart
from aiogram.types import CallbackQuery, Message

from bot.filters import StartWithQuery
from bot.filters import TextCommand
from bot.commands import help_text, help_command, app_command
from bot.gpt.utils import check_subscription
from bot.main_keyboard import create_main_keyboard, send_message
from services import tokenizeService, referralsService

startRouter = Router()

hello_text = """
👋 Привет! Я бот от разработчиков deep.foundation!

🤖 Я готов помочь тебе с любой задачей, просто напиши сообщение или нажми кнопку в меню!

/help - ✨ обзор команд и возможностей
/balance - ⚡️ узнать свой баланс
/referral - 🔗 подробности рефералки

Всё бесплатно, если не платить.
Каждый день твой баланс будет пополняться на сумму от *10 000⚡️* (энергии)!

Если нужно больше:
💎 Можно пополнить баланс Telegram звёздами или банковской картой.
👥 Понравилось? Поделись с друзьями и получи бонус за каждого приглашенного друга! Приводя много друзей ты сможешь пользоваться нейросетями практически безлимитно.

/referral - получай больше с реферальной системой:
*5 000⚡️️* за каждого приглашенного пользователя;
*+500⚡️️* к ежедневному пополнению баланса за каждого друга.

🏠 Если что-то пошло не так или хочешь поделиться вдохновением, напиши в наше сообщество @deepGPT.
"""

ref_text = """
👋 Ты прибыл по реферальной ссылке, чтобы получить награду нужно подписаться на мой канал.
"""

async def handle_referral(message, user_id, ref_user_id):
    result = await referralsService.create_referral(user_id, ref_user_id)
    
    print(result, 'resuuuuult')
    
    # Проверяем, что ref_user_id валиден перед отправкой сообщения
    if not ref_user_id:
        return  # Прерываем выполнение, если ref_user_id отсутствует
    
    if not result or result.get("parent") is None:
        return
    
    # Проверяем, что ref_user_id - число или строка (например, username)
    try:
        chat_id = int(ref_user_id)  # Если ref_user_id должен быть числом
    except (TypeError, ValueError):
        await message.answer("❌ Некорректный реферальный ID.")
        return
    
    await message.answer(text="""
🎉 Вы получили *5 000*⚡️!

/balance - ✨ Узнать баланс
/referral - 🔗 Подробности рефералки
""")

    await message.bot.send_message(
        chat_id=chat_id,  # Используем проверенный chat_id
        text="""
🎉 Добавлен новый реферал! 
Вы получили *5 000*⚡️!
Ваш реферал должен проявить любую активность в боте через 24 часа, чтобы вы получили еще *5 000*⚡️ и +500⚡️️ к ежедневному пополнению баланса.

/balance - ✨ Узнать баланс
/referral - 🔗 Подробности рефералки
"""
    )


async def create_token_if_not_exist(user_id):
    return await tokenizeService.get_tokens(user_id)

@startRouter.message(CommandStart())
async def start(message: types.Message):
    args_match = re.search(r'^/start\s(\S+)', message.text)
    ref_user_id = args_match.group(1) if args_match else None

    # always force sending the keyboard
    keyboard = create_main_keyboard()
    await send_message(message, text=hello_text, reply_markup=keyboard)

    is_subscribe = await check_subscription(message)

    await create_token_if_not_exist(message.from_user.id)

    if not is_subscribe:
        if str(ref_user_id) == str(message.from_user.id):
            return

        await message.answer(
            text=ref_text,
            reply_markup=types.InlineKeyboardMarkup(
                resize_keyboard=True,
                inline_keyboard=[
                    [
                        types.InlineKeyboardButton(text="Подписаться 👊🏻", url="https://t.me/gptDeep"),
                    ],
                    [
                        types.InlineKeyboardButton(text="Проверить ✅",
                                                   callback_data=f"ref-is-subscribe {ref_user_id} {message.from_user.id}"),
                    ]
                ]
            )
        )
        return

    # Уведомляем реферера, что пользователь перешёл по ссылке, но ещё не подписался
    if ref_user_id: # and str(ref_user_id) != str(message.from_user.id):
        try:
            chat_id = int(ref_user_id)
        except (TypeError, ValueError):
            chat_id = None

        if chat_id:
            user_name = message.from_user.username
            user_mention = f"<a href='tg://user?id={message.from_user.id}'>{message.from_user.full_name}</a>"
            await message.bot.send_message(
                chat_id=chat_id,
                text=(
                    f"""
🎉 По вашей реферальной ссылке перешли: @{user_name} ({user_mention}).

Чтобы вашему другу стать рефералом, он должен подписаться на канал @gptDeep.

Как только это произойдёт вы получите <b>5 000</b>⚡️ и <b>+500</b>⚡️️ к ежедневному пополнению баланса.

Если вдруг этого долго не происходит, то возможно вашему другу нужна помощь, попробуйте написать ему в личные сообщения.
"""
                ),
                parse_mode="HTML"
            )
        return

    await handle_referral(message, message.from_user.id, ref_user_id) # TODO: move it to the top (as now it does not work if the user is not subscribed)


@startRouter.callback_query(StartWithQuery("ref-is-subscribe"))
async def handle_ref_is_subscribe_query(callback_query: CallbackQuery):
    ref_user_id = callback_query.data.split(" ")[1]
    user_id = callback_query.data.split(" ")[2]

    is_subscribe = await check_subscription(callback_query.message, user_id)

    if not is_subscribe:
        await callback_query.message.answer(text="Вы не подписались! 😡")
        return

    await handle_referral(callback_query.message, user_id, ref_user_id)


@startRouter.message(TextCommand([help_command(), help_text()]))
async def help_command(message: types.Message):
    await message.answer(text="""
Основной ресурc для доступа к нейросетям - ⚡️ (энергия).
Это универсальный ресурс для всего функционала бота.

Каждая нейросеть тратит разное количество ⚡️.
Количество затраченных ⚡️ зависит от длины истории диалога, моделей нейросетей и объёма ваших вопросов и ответов от нейросети.
Для экономии используйте команду - /clear, чтобы не переполнять историю диалога и не увеличивать расход ⚡️ (энергии)! 
Рекомендуется очищать контекст перед началом обсуждения новой темы. А также если выбранная модель начала отказывать в помощи.

/app - 🔥 Получить ссылку к приложению!
/start - 🔄 Рестарт бота, перезапускает бот, помогает обновить бота до последней версии.
/model - 🛠️ Сменить модель, перезапускает бот, позволяет сменить модель бота.
/system - ⚙️ Системное сообщение, позволяет сменить системное сообщение, чтобы изменить режим взаимодействия с ботом.   
/clear - 🧹 Очистить контекст, помогает забыть боту всю историю.  
/balance - ✨ Баланс, позволяет узнать баланс ⚡️.
/image - 🖼️ Генерация картинки (Midjourney, DALL·E 3, Flux, Stable Diffusion)
/buy - 💎 Пополнить баланс, позволяет пополнить баланс ⚡️.
/referral - 🔗 Получить реферальную ссылку
/suno - 🎵 Генерация музыки (Suno)
/text - Отправить текстовое сообщение
""")


@startRouter.message(TextCommand([app_command()]))
async def app_handler(message: Message):
    await message.answer("""Ссылка на приложение: https://t.me/DeepGPTBot/App""")

File: js/bot/start/router.js
import { helpText, helpCommand, appCommand } from '../commands.js';
import { checkSubscription } from '../gpt/utils.js';
import { createMainKeyboard, sendMessage } from '../main_keyboard.js';
import { tokenizeService, referralsService } from '../../services/index.js';

import { Composer, InlineKeyboard } from 'grammy';
import createDebug from 'debug';
const debug = createDebug('telegram-bot:start_router');

// Greeting text moved to locales/en.yml & ru.yml under key "start.greeting"

const refText = `👋 Ты прибыл по реферальной ссылке, чтобы получить награду нужно подписаться на мой канал.`;

async function handleReferral(ctx, userId, refUserId) {
  const result = await referralsService.createReferral(userId, refUserId);
  console.log(result, 'resuuuuult');
  if (!refUserId) return;
  if (!result || result.parent == null) return;

  let chatId;
  try { chatId = parseInt(refUserId, 10); } catch { chatId = null; }
  if (!chatId) {
    await sendMessage(ctx, '❌ Некорректный реферальный ID.');
    return;
  }

  await sendMessage(ctx, `🎉 Вы получили *5 000*⚡️!\n\n/balance - ✨ Узнать баланс\n/referral - 🔗 Подробности рефералки`);

  await ctx.api.sendMessage(chatId, `🎉 Добавлен новый реферал! \nВы получили *5 000*⚡️!\nВаш реферал должен проявить любую активность в боте через 24 часа, чтобы вы получили еще *5 000*⚡️ и +500⚡️️ к ежедневному пополнению баланса.\n\n/balance - ✨ Узнать баланс\n/referral - 🔗 Подробности рефералки`);
}

async function createTokenIfNotExist(userId) {
  return await tokenizeService.get_tokens(userId);
}

export const startRouter = new Composer();

startRouter.command('start', async (ctx) => {
  debug('Start command handler triggered');
  const refUserId = ctx.match;

  const keyboard = createMainKeyboard();
  const greeting = ctx.t ? ctx.t('start.greeting') : 'Привет!';
  await sendMessage(ctx, { text: greeting, reply_markup: keyboard });

  await createTokenIfNotExist(ctx.from.id);
  const isSubscribe = await checkSubscription(ctx);

  if (refUserId) {
    let chatId = null;
    try { chatId = parseInt(refUserId, 10); } catch {}
    if (chatId) {
      const userName = ctx.from.username;
      const fullName = ctx.from.first_name + (ctx.from.last_name ? ' ' + ctx.from.last_name : '');
      const mention = `<a href='tg://user?id=${ctx.from.id}'>${fullName}</a>`;
      await ctx.api.sendMessage(chatId, `🎉 По вашей реферальной ссылке перешли: @${userName} (${mention}).\n\nЧтобы вашему другу стать вашим рефералом, он должен подписаться на канал @gptDeep.\n\nКак только это произойдёт вы получите <b>5 000</b>⚡️ единоразово и <b>+500</b>⚡️️ к ежедневному пополнению баланса.\n\nЕсли вдруг этого долго не происходит, то возможно вашему другу нужна помощь, <b>попробуйте написать ему в личные сообщения</b>.\nЕсли и это не помогает, то обратитесь в поддержку в сообществе @deepGPT и мы поможем вам разобраться с ситуацией.`, { parse_mode: 'HTML' });
    }
  }

  if (!isSubscribe) {
    if (String(refUserId) !== String(ctx.from.id)) {
      const inlineKb = new InlineKeyboard()
        .url('Подписаться 👊🏻', 'https://t.me/gptDeep').row()
        .text('Проверить ✅', `ref-is-subscribe ${refUserId} ${ctx.from.id}`);
      await sendMessage(ctx, refText, { reply_markup: inlineKb });
    }
    return;
  }

  await handleReferral(ctx, ctx.from.id, refUserId);
});

startRouter.callbackQuery(/ref-is-subscribe (\S+) (\S+)/, async (ctx) => {
  const refUserId = ctx.match[1];
  const userId = ctx.match[2];
  const isSubscribe = await checkSubscription(ctx.message, userId);
  if (!isSubscribe) {
    await sendMessage(ctx, 'Вы не подписались! 😡');
    return;
  }
  await handleReferral(ctx, userId, refUserId);
});

startRouter.hears([helpCommand(), helpText()], async (ctx) => {
  await sendMessage(ctx, `Основной ресурc для доступа к нейросетей - ⚡️ (энергия).\nЭто универсальный ресурс для всего функционала бота.\n\nКаждая нейросеть тратит разное количество ⚡️.\nКоличество затраченных ⚡️ зависит от длины истории диалога, моделей нейросетей и объёма ваших вопросов и ответов от нейросети.\nДля экономии используйте команду - /clear, чтобы не переполнять историю диалога и не увеличивать расход ⚡️ (энергии)! \nРекомендуется очищать контекст перед началом обсуждения новой темы. А также если выбранная модель начала отказывать в помощи.\n\n/app - 🔥 Получить ссылку к приложению!\n/start - 🔄 Рестарт бота, перезапускает бот, помогает обновить бота до последней версии.\n/model - 🛠️ Сменить модель, перезапускает бот, позволяет сменить модель бота.\n/system - ⚙️ Системное сообщение, позволяет сменить системное сообщение, чтобы изменить режим взаимодействия с ботом.   \n/clear - 🧹 Очистить контекст, помогает забыть боту всю историю.  \n/balance - ✨ Баланс, позволяет узнать баланс ⚡️.\n/image - 🖼️ Генерация картинки (Midjourney, DALL·E 3, Flux, Stable Diffusion)\n/buy - 💎 Пополнить баланс, позволяет пополнить баланс ⚡️.\n/referral - 🔗 Получить реферальную ссылку\n/suno - 🎵 Генерация музыки (Suno)\n/text - Отправить текстовое сообщение`);
});

startRouter.hears(appCommand(), async (ctx) => {
  await sendMessage(ctx, 'Ссылка на приложение: https://t.me/DeepGPTBot/App');
});
